// @ibm-cloud/cloudant/index.js
"use strict";
/**
 * © Copyright IBM Corporation 2020, 2025. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PagerType = exports.Pagination = exports.Stream = exports.ChangesFollower = exports.CouchdbSessionAuthenticator = exports.IamAuthenticator = exports.BasicAuthenticator = exports.CloudantV1 = void 0;
/**
 * @module cloudant-node-sdk
 */
const v1_1 = __importDefault(require("./cloudant/v1"));
Object.defineProperty(exports, "CloudantV1", { enumerable: true, get: function () { return v1_1.default; } });
var auth_1 = require("./auth");
Object.defineProperty(exports, "BasicAuthenticator", { enumerable: true, get: function () { return auth_1.BasicAuthenticator; } });
Object.defineProperty(exports, "IamAuthenticator", { enumerable: true, get: function () { return auth_1.IamAuthenticator; } });
Object.defineProperty(exports, "CouchdbSessionAuthenticator", { enumerable: true, get: function () { return auth_1.CouchdbSessionAuthenticator; } });
var changesFollower_1 = require("./cloudant/features/changesFollower");
Object.defineProperty(exports, "ChangesFollower", { enumerable: true, get: function () { return changesFollower_1.ChangesFollower; } });
var stream_1 = require("./cloudant/features/stream");
Object.defineProperty(exports, "Stream", { enumerable: true, get: function () { return stream_1.Stream; } });
var pagination_1 = require("./cloudant/features/pagination/pagination");
Object.defineProperty(exports, "Pagination", { enumerable: true, get: function () { return pagination_1.Pagination; } });
Object.defineProperty(exports, "PagerType", { enumerable: true, get: function () { return pagination_1.PagerType; } });
